### ISSUE TITLE
why esbuild plugin use js

### ISSUE THREAD

**Issue by stoneyallen** [evanw/esbuild#1344]
It is well known that the reason esbuild is fast is because of the go-based environment and not having to start node platform services. However, if esbuild's plugins are written in js, it increases the node startup time and runs counter to the original design intent??

**Comment by EMH333** [comment]
I respectfully disagree that JS plugins are out of the original design intent/bad for esbuild for several reasons (I am NOT the author of esbuild, this is my opinion only):

1. They are totally optional. You can run esbuild without running JS and it will work just fine. There also is a golang based plugin API which allows for plugins while avoiding JS
2. Because they aren't some after-thought, the code used to run them is performant. If I recall they are run in parallel if possible and significant thought has been put into reducing the overhead of the `native <-> node` communication
3. They provide the avenue for a ton of use cases. For example, the plugin I maintain (https://github.com/EMH333/esbuild-svelte) allows `.svelte` files to be compiled and optimized without resorting to some other JS bundler. It performs well and the Svelte compile time is going to be the same regardless of bundler. **I'm not going to try and port all of Svelte to go**, and there are many JS plugins which have a similar setup.
4. They reduce scope creep in esbuild. If it can be done in a plugin and isn't something widely used then it can go in a plugin instead of esbuild. There are a ton of JS plugins (see https://github.com/esbuild/community-plugins) which add additional features that are way, way out of scope for esbuild as a bundler but can be surprisingly useful in certain situations.

I find JS plugins to add incredible value to esbuild even if they run slightly slower.

I'm curious to what degree this impacts you, even with a "full load" of plugins I still find esbuild to be much faster than any other bundler I have tried.

**Comment by evanw** [comment]
JS plugins exist because there are many use cases where you need to bind to a library that’s written in JS. Using JS is the only option in these cases. 

There is also a Go plugin API that is equivalent to the JS plugin API. If you don’t want to use JS plugins, then just don’t use them. There should be nothing that forces you to use JS plugins. 

I’m closing this issue as “by design.”
